pkg/redirect/redirect.go

```diff
@@ -0,0 +1,324 @@
+				} else {
+					vv = strings.Replace(vv, `realm="https://ghcr.io/`, fmt.Sprintf(`realm="https://%s/`, req.Host), 1)
+				}
+				log.Println("=== CHANGED: Www-Authenticate:", vv)
+			}
+			resp.Header().Add(k, vv)
+		}
+	}
+	resp.WriteHeader(back.StatusCode)
+	if _, err := io.Copy(resp, back.Body); err != nil {
+		logger.Errorf("Error copying response body: %v", err)
+	}
+}
+
+func (rdr redirect) token(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.FromContext(ctx)
+
+	vals := r.URL.Query()
+	if rdr.prefix != "" {
+		scope := vals.Get("scope")
+		scope = strings.Replace(scope, rdr.prefix+"/", "", 1)
+		vals.Set("scope", scope)
+	}
+	if rdr.repo != "" {
+		scope := vals.Get("scope")
+		scope = strings.Replace(scope, "repository:", "repository:"+rdr.repo+"/", 1)
+		vals.Set("scope", scope)
+	}
+
+	var url string
+	if rdr.host == "gcr.io" {
+		url = "https://gcr.io/v2/token?" + vals.Encode()
+	} else {
+		url = "https://ghcr.io/token?" + vals.Encode()
+	}
+
+	req, _ := http.NewRequest(r.Method, url, nil)
+	req.Header = r.Header.Clone()
+
+	logger.Infow("sending request",
+		"method", req.Method,
+		"url", req.URL.String(),
+		"header", redact(req.Header))
+	w.Header().Set("X-Redirected", req.URL.String())
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		logger.Errorf("Error sending request: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	defer resp.Body.Close()
+
+	logger.Infow("got response",
+		"method", req.Method,
+		"url", req.URL.String(),
+		"status", resp.Status,
+		"header", redact(resp.Header))
+
+	for k, v := range resp.Header {
+		for _, vv := range v {
+			w.Header().Add(k, vv)
+		}
+	}
+	w.WriteHeader(resp.StatusCode)
+	if _, err := io.Copy(w, resp.Body); err != nil {
+		logger.Errorf("Error copying response body: %v", err)
+	}
+}
+
+func (rdr redirect) proxy(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := logging.FromContext(ctx)
+
+	var url string
+	if rdr.host == "gcr.io" {
+		url = "https://gcr.io/v2/"
+	} else {
+		url = "https://ghcr.io/v2/"
+	}
+	if rdr.repo != "" {
+		url += rdr.repo + "/"
+	}
+
+	path := strings.TrimPrefix(r.URL.Path, "/v2/")
+	if rdr.prefix != "" {
+		// Trim the prefix, if any.
+		// If the path didn't have the prefix, that's fine for now too.
+		path = strings.TrimPrefix(path, rdr.prefix+"/")
+	}
+
+	url += path
+	if query := r.URL.Query().Encode(); query != "" {
+		url += "?" + query
+	}
+	req, _ := http.NewRequest(r.Method, url, nil)
+	req.Header = r.Header.Clone()
+
+	// If the request is coming in without auth, get some auth.
+	// This is useful for testing, but should never happen in real life.
+	// Actually, containerd seems to make unauthenticated HEAD requests before
+	// hitting /v2/, so this might be load-bearing.
+	if req.Header.Get("Authorization") == "" {
+		t, resp, err := rdr.getToken(r)
```

Comment:
if not supposed to happen, maybe worth a log line here like "No Authorization header detected, attempting to fetch token"